Type mismatching cased by duplicate associated type resolution

Apologize that I could not give any code example right now. I'm working on creating one but it seems tricky. 

The compiler is complaining about `[E0308]: mismatched types` where it said it's expecting for a trait `SomeTrait<Apple=SomeApple, Apple=SomeApple, Banana=SomeBanana, Banana=SomeBanana>` but got `SomeTrait<Apple=SomeAPple, Banana=SomeBanana>`. Only some of the associated types in that trait are repeated once and there are few others are not repeated.

The example I gave in #59324 is an intermediate result of my attempt to create a minimal reproducible code to this issue. The type mismatch happened at the `with_factory` method when resolving concrete type for `ThriftService`.

The error happened when I'm doing the update from 1.32.0 to 1.33.0 so it's clearly a regression. 

Also, please let me know if you have any suggestion on bypassing this problem or tips on re-creating the problematic code.

Thanks.

Okay, I'm able to create a small example demonstrating this problem:

```

use std::marker::PhantomData;

use bytes::Bytes; // 0.4.11;

trait Service {
    type Request;
    type Response;
}

trait Framing {
    type Request;
    type Response;
}

trait HttpService<B>: Service<Request = B::Request, Response = B::Response>
where
    B: Framing + Send + 'static,
{
    type Handler;
}

pub struct Http2Transport;
impl Framing for Http2Transport {
    type Request = Bytes;
    type Response = Bytes;
}

type BoxService<H> = Box<
    HttpService<Http2Transport, Handler = H, Request = Bytes, Response = Bytes> + Send + 'static,
>;

fn build_server<F, SVC, H>(factory: F)
where
    F: FnOnce() -> SVC + Send + Sync + Clone + 'static,
    SVC: Fn(i32) -> Result<BoxService<H>, std::io::Error>,
    H: 'static,
{
}

pub trait MyAwesomeService: Send + Sync + 'static {}

struct MyServiceImpl;
impl MyAwesomeService for MyServiceImpl {}

trait ProtocolFactory {
    type Frame: Framing;
}

pub struct Http2ProtocolFactory<F> {
    _phantom: PhantomData<F>,
}

impl<F> ProtocolFactory for Http2ProtocolFactory<F>
where
    F: Framing,
{
    type Frame = F;
}

pub struct ServiceProcessor<PF, H> {
    service: H,
    _phantom: PhantomData<(PF, H)>,
}

impl<PF, H> ServiceProcessor<PF, H>
where
    PF: ProtocolFactory + 'static,
    H: MyAwesomeService,
{
    pub fn new(service: H) -> Self {
        Self {
            service,
            _phantom: PhantomData,
        }
    }
}

impl<PF, H> Service for ServiceProcessor<PF, H>
where
    PF: ProtocolFactory + 'static,
    PF::Frame: Send + 'static,
    H: MyAwesomeService,
{
    type Request = <<PF as ProtocolFactory>::Frame as Framing>::Request;
    type Response = <<PF as ProtocolFactory>::Frame as Framing>::Response;
}

impl<PF, H> HttpService<PF::Frame> for ServiceProcessor<PF, H>
where
    PF: ProtocolFactory + 'static,
    PF::Frame: Send + 'static,
    H: MyAwesomeService,
{
    type Handler = H;
}

fn make_server<F, H>(
    proto: i32,
    handler: H,
) -> Result<
    Box<
        HttpService<
                F,
                Handler = H,
                Request = <F as Framing>::Request,
                Response = <F as Framing>::Response,
            > + Send
            + 'static,
    >,
    std::io::Error,
>
where
    F: Framing + Send + 'static,
    H: MyAwesomeService,
{
    Ok(Box::new(
        ServiceProcessor::<Http2ProtocolFactory<F>, H>::new(handler),
    ))
}

fn main() {
    build_server(|| |i: i32| make_server(i, MyServiceImpl {}))
}
```

This code compiles in 1.32.0 but not in 1.33.0. Tested locally:

```
type-resolution $ cargo version
cargo 1.32.0 (8610973aa 2019-01-02)
type-resolution $ cargo build
...
    Finished dev [unoptimized + debuginfo] target(s) in 3.36s
type-resolution $ rustup default 1.33.0
...
  1.33.0-x86_64-apple-darwin installed - rustc 1.33.0 (2aa4c46cf 2019-02-28)
type-resolution $ cargo version
cargo 1.33.0 (f099fe94b 2019-02-12)
type-resolution $ cargo build
...

error[E0308]: mismatched types
   --> src/main.rs:124:30
    |
124 |     build_server(|| |i: i32| make_server(i, MyServiceImpl {}))
    |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected trait `HttpService<Http2Transport, Response=bytes::bytes::Bytes, Response=bytes::bytes::Bytes, Request=bytes::bytes::Bytes, Request=bytes::bytes::Bytes, Handler=_> + std::marker::Send`, found trait `HttpService<_, Response=_, Request=_, Handler=MyServiceImpl> + std::marker::Send`
    |
    = note: expected type `std::result::Result<std::boxed::Box<(dyn HttpService<Http2Transport, Response=bytes::bytes::Bytes, Response=bytes::bytes::Bytes, Request=bytes::bytes::Bytes, Request=bytes::bytes::Bytes, Handler=_> + std::marker::Send + 'static)>, _>`
               found type `std::result::Result<std::boxed::Box<(dyn HttpService<_, Response=_, Request=_, Handler=MyServiceImpl> + std::marker::Send + 'static)>, _>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
error: Could not compile `type-resolution`.
```

Not sure, if it is the smallest possible code, but it's pretty darn good

```rust
trait Service {
    type S;
}

trait Framing {
    type F;
}

impl Framing for () {
    type F = ();
}

trait HttpService<F: Framing>: Service<S = F::F> {}

type BoxService = Box<dyn HttpService<(), S = ()>>;

fn build_server<F: FnOnce() -> BoxService>(_: F) {}

fn make_server<F: Framing>() -> Box<dyn HttpService<F, S = F::F>> {
    unimplemented!()
}

fn main() {
    build_server(|| make_server())
}
```

```text
1.32.0    Checking foo v0.1.0 (/tmp/tmp.eBIGDgl0Zm/foo)
    Finished dev [unoptimized + debuginfo] target(s) in 0.30s
1.33.0    Checking foo v0.1.0 (/tmp/tmp.eBIGDgl0Zm/foo)
error[E0308]: mismatched types
  --> src/main.rs:24:21
   |
24 |     build_server(|| make_server())
   |                     ^^^^^^^^^^^^^ expected trait `HttpService<(), S=(), S=()>`, found trait `HttpService<_, S=_>`
   |
   = note: expected type `std::boxed::Box<(dyn HttpService<(), S=(), S=()> + 'static)>`
              found type `std::boxed::Box<(dyn HttpService<_, S=_> + 'static)>`

error: aborting due to previous error
```

#73485 would solve this and incur no perf regression, but I noticed that with the following code

```rust
trait Service {
    type S;
}

trait Framing {
    type F;
}

impl Framing for () {
    type F = ();
}

impl Framing for u32 {
    type F = u32;
}

trait HttpService<F: Framing>: Service<S = F::F> {}

type BoxService = Box<dyn HttpService<u32, S = ()>>;

fn build_server<F: FnOnce() -> BoxService>(_: F) {}

fn make_server<F: Framing>() -> Box<dyn HttpService<F, S = F::F>> {
    unimplemented!()
}

fn main() {
    build_server(|| make_server())
}
```

we don't get any specialized error, we would still be getting what we get now

```
error[E0308]: mismatched types
  --> src/main.rs:28:21
   |
28 |     build_server(|| make_server())
   |                     ^^^^^^^^^^^^^ expected trait `HttpService<u32, S = (), S = u32>`, found trait `HttpService<_, S = _>`
   |
   = note: expected struct `std::boxed::Box<(dyn HttpService<u32, S = (), S = u32> + 'static)>`
              found struct `std::boxed::Box<(dyn HttpService<_, S = _> + 'static)>`
```

but we should still handle the case of multiple predicates having conflicting obligations with a clearer error. This is a tangentially related problem to the raised case of multiple predicates with the same obligation causing the return of `ExistentialMismatch` instead of succeeding.

--- a/src/librustc_middle/ty/relate.rs
+++ b/src/librustc_middle/ty/relate.rs
@@ -617,12 +617,22 @@ impl<'tcx> Relate<'tcx> for &'tcx ty::List<ty::ExistentialPredicate<'tcx>> {
         a: &Self,
         b: &Self,
     ) -> RelateResult<'tcx, Self> {
-        if a.len() != b.len() {
+        let tcx = relation.tcx();
+
+        // FIXME: this is wasteful, but want to do a perf run to see how slow it is.
+        // We need to perform this deduplication as we sometimes generate duplicate projections
+        // in `a`.
+        let mut a_v: Vec<_> = a.into_iter().collect();
+        let mut b_v: Vec<_> = b.into_iter().collect();
+        a_v.sort_by(|a, b| a.stable_cmp(tcx, b));
+        a_v.dedup();
+        b_v.sort_by(|a, b| a.stable_cmp(tcx, b));
+        b_v.dedup();
+        if a_v.len() != b_v.len() {
             return Err(TypeError::ExistentialMismatch(expected_found(relation, a, b)));
         }
 
-        let tcx = relation.tcx();
-        let v = a.iter().zip(b.iter()).map(|(ep_a, ep_b)| {
+        let v = a_v.into_iter().zip(b_v.into_iter()).map(|(ep_a, ep_b)| {
             use crate::ty::ExistentialPredicate::*;
             match (ep_a, ep_b) {
                 (Trait(ref a), Trait(ref b)) => Ok(Trait(relation.relate(a, b)?)),

--- a/src/test/ui/issues/issue-59326.rs
+++ b/src/test/ui/issues/issue-59326.rs
@@ -1,0 +1,26 @@
+// check-pass
+trait Service {
+    type S;
+}
+
+trait Framing {
+    type F;
+}
+
+impl Framing for () {
+    type F = ();
+}
+
+trait HttpService<F: Framing>: Service<S = F::F> {}
+
+type BoxService = Box<dyn HttpService<(), S = ()>>;
+
+fn build_server<F: FnOnce() -> BoxService>(_: F) {}
+
+fn make_server<F: Framing>() -> Box<dyn HttpService<F, S = F::F>> {
+    unimplemented!()
+}
+
+fn main() {
+    build_server(|| make_server())
+}

